internal/pkg/agent/application/upgrade/cleanup_test.go

```diff
@@ -0,0 +1,53 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
+
+	"github.com/stretchr/testify/require"
+)
+
+func setupDir(t *testing.T) {
+	t.Helper()
+	dir := t.TempDir()
+	paths.SetDownloads(dir)
+
+	err := os.WriteFile(filepath.Join(dir, "test-8.3.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-8.4.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-8.5.0-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+	err = os.WriteFile(filepath.Join(dir, "test-hash-file"), []byte("hello, world!"), 0600)
+	require.NoError(t, err)
+}
```

Comment:
A bit late to the discussion here, but there is an `FS Interface` at https://pkg.go.dev/io/fs#FS, this allow you mock everything in memory and not have to deal with the weird syncing issues on drives.